Unique Network Usage (#237)
BCDevOps/OpenShift4-RollOut: docs/Cluster-Internal-networks.md

# Cluster Internal Network design

> This document contains the information used to decide whether to provision separate (unique) networks for each OpenShift 4 cluster.  The current design requires routeable (public) IP addresses only at the F5 ingress/egress points for each cluster, and we are not looking to change the design of using non-routeable IP addresses for the following networks.  The decision in question is whether we can re-use the same private network ranges for multiple clusters.  This decision supports a critical cluster component, which if changed, would require a full re-installation of a given cluster.

Author: ```suggestion
This document contains the information used to decide whether to provision separate (unique) networks for each OpenShift 4 cluster.  The current design requires routeable (public) IP addresses only at the F5 ingress/egress points for each cluster, and we are not looking to change the design of using private (non-internet routable) IP addresses for the rest of the networks.  The decision in question is whether we can re-use the same private network ranges for multiple clusters.  This decision supports a critical cluster component, which if changed, would require a full re-installation of a given cluster.
```

> This document contains the information used to decide whether to provision separate (unique) networks for each OpenShift 4 cluster.  The current design requires routeable (public) IP addresses only at the F5 ingress/egress points for each cluster, and we are not looking to change the design of using private (non-internet routable) IP addresses for the rest of the networks.  The decision in question is whether we can re-use the same private network ranges for multiple clusters.  This decision supports a critical cluster component, which if changed, would require a full re-installation of a given cluster.

Reviewer: Are we talking about the host IPs of the nodes, or the Service/Cluster SDN networks of OpenShift?

Reviewer: Nevermind, this is made clear below

> This document contains the information used to decide whether to provision separate (unique) networks for each OpenShift 4 cluster.  The current design requires routeable (public) IP addresses only at the F5 ingress/egress points for each cluster, and we are not looking to change the design of using private (non-internet routable) IP addresses for the rest of the networks.  The decision in question is whether we can re-use the same private network ranges for multiple clusters.  This decision supports a critical cluster component, which if changed, would require a full re-installation of a given cluster.

Author: ```suggestion
This document contains the information used to decide whether to provision separate (unique) networks for each OpenShift 4 cluster.  The current design requires public IP addresses at the F5 ingress/egress points for each cluster, and private IP addresses for the rest of the networks.  The decision in question is whether we can re-use the same private network ranges for multiple clusters.  This decision supports critical cluster network components, which if changed, would require a full re-installation of a given cluster.
```

## Background

With each OpenShift cluster, internal networks are used for in-cluster communications.  These networks are private networks that rely on ingress and egress Network Address Translation.  If a network does not require external direct routing/exposure, then we can save on IP network reservations by re-using these private networks.  The following networks are the potential private networks for OCP 4:

- Host/Node Network - primary physical interface for each physical host in a cluster (outgoing NAT IP address for all pods on a node)
- Storage Network - connecting NAS storage (NetApp) to each of the hosts within a cluster.
- Pod Network - IP addresses assigned to each pod.
- Service Network - IP addresses assigned to services.

## Network usage

### Host/Node Network

> Host IP addresses are used to NAT each hosted PODs outgoing traffic, as well as host specific node/cluster related services.  Incoming application traffic will not be directed at a host IP address.  While we are planning to use outgoing NAT on the F5, there are situations where we may want to be able to direct route to these host IP addresses from inside the data-centres.  Current OCP3 leverages public IP addresses for all host/nodes so a pre-build list of services that would require direct access has not been described/tested yet.

Reviewer: > there are situations where we may want to be able to direct route to these host IP addresses from inside the data-centres.

I can't think of any examples. Though the OCP3 cluster uses public IPs, they are all still behind a firewall. There are no firewall rules that I know of to allow incoming traffic to the host vlan.

Author: Are you asking to change the recommendation from unique host IP ranges to overlapping host IP ranges?

Reviewer: I'm asking for non-routable ranges. Unique should be no issue.

Author: cool, I think I'm having issues with clarity in the doc, and any suggestions on wording are appreciated.

by routable, I do not mean public ranges, just unique within BCGov to allow for internal routing.  Can you think of a better way to word that? or should I just remove "routable" and stick with private and unique/reserved?

Also, I have not seen any diagrams of your internal network structure/setup, so it's all conjecture from my perspective on how you've setup the storage/management/primary/backup network interfaces.  If you have additional network diagrams/documentation you can share there, that would also be appreciated.

Reviewer: > just unique within BCGov to allow for internal routing

Where do you need to route to?

> Host IP addresses are used to NAT each hosted PODs outgoing traffic, as well as host specific node/cluster related services.  Incoming application traffic will not be directed at a host IP address.  While we are planning to use outgoing NAT on the F5, there are situations where we may want to be able to direct route to these host IP addresses from inside the data-centres.  Current OCP3 leverages public IP addresses for all host/nodes so a pre-build list of services that would require direct access has not been described/tested yet.

Author: ```suggestion
Host IP addresses are used to NAT each hosted PODs outgoing traffic, as well as host specific node/cluster related services.  Incoming application traffic will not be directed at a host IP address.  While we are planning to use outgoing NAT on the F5, there may be situations in the future where we need to be able to direct route to these host IP addresses from other BC Gov networks (eg: future Aporeto design potential for bridging zone access may need route-able endpoints).

Changing the Host/Node network in the future would require a full cluster rebuild or migration.
```

> A separate (unique) Host/Node network for each cluster is recommended.

Reviewer 2: A separate unique private routable Host/Node network for each cluster is recommended.

> A separate (unique) Host/Node network for each cluster is recommended.

Author: Would this capture your suggestion @pfischer8989 ?
```suggestion
A separate, unique, private and routable Host/Node network for each cluster is recommended.
```

Reviewer: What needs to route to these IPs?

> A separate, unique, private and routable Host/Node network for each cluster is recommended.

Author: ```suggestion
A private and route-able Host/Node network for each cluster is recommended.
```

### Storage Network

> This network will be leveraged for high speed communication between the NetApp appliance and the cluster Hosts/Nodes.  While no other access is planned for this network, there are still design elements to identify (eg: backup network integration, etc) that may benefit from a unique network assignment.

Author: ```suggestion
This network will be leveraged for high speed communication between the NetApp appliance and the cluster Hosts/Nodes.  No other access is planned for a cluster's storage network, storage is not shared between clusters, and backup service integration will leverage a separate backup interface as determined by the hosting service.
```

> A separate (unique) Storage network for each cluster is recommended.

Reviewer 2: A separate unique private (non-routable??)  Storage network for each cluster is recommended.

I am not sure if the Gov does non-routable storage network blocks..... I have certainly done this in the past depending on the requirements of the cluster.

> A separate (unique) Storage network for each cluster is recommended.

Author: My initial recommendation for private routing to this storage network is to support integrating this storage with the BCGov Hosted backup service.  (I am not sure how many network interfaces they're configuring on the netapp devices for management/backup/storage access)
Does the following capture your change comment?
```suggestion
A separate, unique, private Storage network for each cluster is recommended.  With private routing to be determined by Backup service requirements.
```

Reviewer: Backups will be done via a separate vlan and interfaces. Backup team wouldn't want us using their vlans for our in-cluster traffic.

We are using private non-routable vlans today with the NetApp VMs without issues.

> A separate, unique, private Storage network for each cluster is recommended.  With private routing to be determined by Backup service requirements.

Author: ```suggestion
A private and non-routable Storage network for each cluster is recommended.
```

### Cluster Pod and Service Networks

These networks are used by internal pod-pod communication while external cluster access (both incoming and outgoing) is routed through the host/node IP address or a routed ingress IP.  This makes these networks good candidates for re-using network ranges.

A known technology that will be impacted by this decision is the Istio Service Mesh implementation.  The internal pod and service networks are used for service discovery within a Mesh backplane.  One design for a multi-cluster deployment uses a single (shared) Istio backplane.  This design requires uniquely identifiable pod endpoints.  A second design uses separate Istio backplanes for each cluster, and replication between backplanes provides cross-cluster service discovery.  This design would NOT require uniquely identifiable pod endpoints between clusters.

> Choosing to re-use the internal cluster networks (for pods and services) will remove the option to stretch an Istio Backplane across multiple clusters.  Requesting additional private networks for each production cluster will require requests for reserved network ranges from hosting services.

Reviewer: How does Istio communicate between clusters? Does each cluster just expose an API endpoint via a route and talk over that?

Author: One Istio installation option is to have separate backplane installation (as described here) where the service discover/routing will be done through an Istio Gateway.  The other is where the Istio backplane is stretched across the clusters and service discovery/routing is done directly against the cluster services/pods.  At this point the direction for multi-cluster Istio with Redhat is still in development, and the recommendation reflects a balance between keeping our options open, and reducing our unique network reservation requirements.

Reviewer: > service discovery/routing is done directly against the cluster services/pods

What network is this done over? How does one cluster communicate to the other? Is this why the Host Network needs to be routable?

The current production architecture defines a *Silver* platform that includes a single cluster, and *Gold* platform that is made up of 2 clusters in separate physical regions.

> The recommendation is to leverage unique cluster pod/service networks within each region.  This would allow the option of a single Istio Backplane across the Gold clusters in the future, while not demanding unique pod/service networks in every cluster.  This would allow partial re-use of reserved networks while not blocking a future design pattern we may need to leverage.

Reviewer: I think we should also cover how big of a network this needs to be.

https://docs.openshift.com/container-platform/4.4/installing/installing_bare_metal/installing-bare-metal.html#installation-bare-metal-config-yaml_installing-bare-metal

`hostPrefix` determines how many pods you can have on a host. This can't be changed later. Red Hat recommends 23 to allow for 512 pods per host as is allowed in OCP 4.4.

`hostPrefix` times the maximum number of nodes in the cluster will give the full CIDR size needed.

It's my understanding that both `clusterNetwork` and `serviceNetwork` should be the same size.

Author: ```suggestion
The recommendation is to leverage private non-routable cluster pod/service networks within each region.  This would enable the possibility for a single Istio Backplane across the Gold clusters in the future, while not demanding separate non-routable pod/service networks in every cluster.
```
